examples/powercli/nsx-t_tag-sync/handler/script.ps1

```diff
@@ -0,0 +1,160 @@
+#   NSX-T 3.0.2
+#
+# Notes:
+#   Machine names in a vCenter instance must be unique: 
+#       Script has no way, based on limited specificity of vSphere event message of the types above, to discern correct machine other than by name.
+#           Lots of issues here as the name may not be unique in the cluster
+#           Intention is to fix this as vSphere event messages evolve.
+#       Event data has a number of odd charachters, such as new lines, that must be accomodated to get the machine name.  
+#           accomodation efforts to handle naming are ongoing as errata is reported.
+#
+# Planned enhancements (script today is minimally viable):
+#   Updates based on vSphere event changes
+#   Hopefully, better vSphere and NSX-T integration will render this script obsolete.
+
+# Process function Secrets passed in
+$SECRETS_FILE = "/var/openfaas/secrets/nsx-secrets"
+$SECRETS_CONFIG = (Get-Content -Raw -Path $SECRETS_FILE | ConvertFrom-Json)
+
+# Test if PowerCLI Module is installed, install if not
+if(Get-Module -ListAvailable -Name VMware.VimAutomation.Core) {
+    Write-Host "Module exists"
+} else {
+    Write-Host "Module does not exist"
+    Install-Package -Name VMware.VimAutomation.Core
+}
+Set-PowerCLIConfiguration -InvalidCertificateAction Ignore  -DisplayDeprecationWarnings $false -ParticipateInCeip $false -Confirm:$false | Out-Null
+
+# Process payload sent from vCenter Server Event
+$json = $args | ConvertFrom-Json
+
+if($env:function_debug -eq "true") {
+    Write-Host "DEBUG: json=`"$($json | Format-List | Out-String)`""
+}
+
+# Process payload sent from vCenter Server Event
+$vcenter = ($json.source -replace "https://","" -replace "/sdk","")
+
+# Pull VM name from event message text and set it to variable.  
+$separator = "object"
+$FullFormattedMessage = $json.data.FullFormattedMessage
+if($env:function_debug -eq "true") {
+    write-host "FullFormattedMessage RAW="$FullFormattedMessage
+}
+$FullFormattedMessage.replace([Environment]::NewLine," ")
+if($env:function_debug -eq "true") {
+    write-host "FullFormattedMessage minus NewLine="$FullFormattedMessage
+}
+$pos = $FullFormattedMessage.IndexOf($separator)
+$rightPart = $FullFormattedMessage.Substring($pos+1)
+if($env:function_debug -eq "true") {
+    $leftPart = $FullFormattedMessage.Substring(0, $pos)
+    write-host "FullFormattedMessage leftPart="$leftPart
+    write-host "FullFormattedMessage rightPart="$rightPart
+}
+$pos = $rightPart.replace("bject","")
+$FormattedMessage = $pos.replace([Environment]::NewLine," ")
+if($env:function_debug -eq "true") {
+    write-host "FullFormattedMessage Split="$FullFormattedMessage
+}
+$FormattedMessage = $FormattedMessage.trim()
+if($env:function_debug -eq "true") {
+    write-host "FullFormattedMessage Complete="$FullFormattedMessage
+}
+$vm = $FormattedMessage
+
+if($vmMoRef -eq "" -or $vm -eq "") {
+    Write-Host "Unable to retrieve VM Object from Event payload, please ensure Event contains VM result"
+    exit
+}
```

Comment:
There's actually an easier way to retrieve the VM/Object that the tag was applied to. 

Here's what the "Attach" Event looks like from PowerCLI:
```
EventTypeId          : com.vmware.cis.tagging.attach
Severity             : info
Message              :
Arguments            : {Tag, Object, User}
ObjectId             :
ObjectType           :
ObjectName           :
Fault                :
Key                  : 530152
ChainId              : 530152
CreatedTime          : 2/3/2021 10:22:15 AM
UserName             : VSPHERE.LOCAL\Administrator
Datacenter           :
ComputeResource      :
Host                 :
Vm                   :
Ds                   :
Net                  :
Dvs                  :
FullFormattedMessage : User VSPHERE.LOCAL\Administrator attached tag Test to object VMTX-VM
ChangeTag            :
```

Notice there's `arguments` property which contains array of key/value, one of which is `Object` key

```
 > $tagEvent.Arguments

Key    Value
---    -----
Tag    Test
Object VMTX-VM
User   VSPHERE.LOCAL\Administrator
```

To filter out the Object, you can use the following snippet which looks for `Object` keyword:
```
 > ($tagEvent.Arguments | where {$_.key -eq "Object"}).Value
VMTX-VM
```

Reply:
Awesome!  I'll make the change.